rclcpp_action/test/test_server.cpp

```diff
@@ -1234,10 +1234,13 @@ class TestDeadlockServer : public TestServer
         this->TryLockFor(lock, std::chrono::milliseconds(1000));
         return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
       },
-      [this](std::shared_ptr<GoalHandle>) {
+      [this](std::shared_ptr<GoalHandle> handle) {
         // instead of making a deadlock, check if it can acquire the lock in a second
         std::unique_lock<std::recursive_timed_mutex> lock(server_mutex_, std::defer_lock);
         this->TryLockFor(lock, std::chrono::milliseconds(1000));
+        if (!handle->is_active()) {
+          return rclcpp_action::CancelResponse::REJECT;
+        }
```

Comment:
Should we have a TODO here mentioning the issue @fujitatomoya created?

```suggestion
        # TODO(KavenYau): this check may become obsolete with https://github.com/ros2/rclcpp/issues/1599
        if (!handle->is_active()) {
          return rclcpp_action::CancelResponse::REJECT;
        }
```

Or is the thinking that we'll need it always?

Reply:
I think we should patch things so that users don't have to check if the goal is active. IIUC, there's still potentially a race doing this in general, since the goal state may change between checking `is_active()` and returning from the user-callback.